framework/TSA/ARMA.py

```diff
@@ -0,0 +1,251 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+  AutoRegressive Moving Average time series analysis
+"""
+import copy
+import collections
+import numpy as np
+import scipy as sp
+
+import Decorators
+
+from utils import InputData, InputTypes, randomUtils, xmlUtils, mathUtils, importerUtils
+statsmodels = importerUtils.importModuleLazy('statsmodels', globals())
+
+import Distributions
+from .TimeSeriesAnalyzer import TimeSeriesAnalyzer
+
+
+# utility methods
+class ARMA(TimeSeriesAnalyzer):
+  r"""
+    AutoRegressive Moving Average time series analyzer algorithm
+  """
+  # class attribute
+  ## define the clusterable features for this trainer.
+  _features = [] # TODO
+
+  @classmethod
+  def getInputSpecification(cls):
+    """
+      Method to get a reference to a class that specifies the input data for
+      class cls.
+      @ Out, inputSpecification, InputData.ParameterInput, class to use for
+        specifying input of cls.
+    """
+    specs = super(ARMA, cls).getInputSpecification()
+    specs.name = 'arma' # NOTE lowercase because ARMA already has Fourier and no way to resolve right now
+    specs.description = r"""TimeSeriesAnalysis algorithm for determining the stochastic
+                            characteristics of signal with time-invariant variance"""
+    specs.addParam('reduce_memory', param_type=InputTypes.BoolType, required=False,
+                   descr=r"""activates a lower memory usage ARMA training. This does tend to result
+                   in a slightly slower training time, at the benefit of lower memory usage. For
+                   example, in one 1000-length history test, low memory reduced memory usage by 2.3
+                   MiB, but increased training time by 0.4 seconds. No change in results has been
+                   observed switching between modes. Note that the ARMA must be
+                   retrained to change this property; it cannot be applied to serialized ARMAs.
+                   \default{False}""")
+    specs.addSub(InputData.parameterInputFactory('SignalLag', contentType=InputTypes.FloatType,
+                 descr=r"""the number of terms in the AutoRegressive term to retain in the
+                           regression; "P" in literature."""))
+    specs.addSub(InputData.parameterInputFactory('NoiseLag', contentType=InputTypes.FloatType,
+                 descr=r"""the number of terms in the Moving Average term to retain in the
+                           regression; "Q" in literature."""))
```

Comment:
What do you think of adding _"Represented as "Q" in literature."_ or _"Typically Represented as "Q" in literature."_ Just to be a little more explicit?

Reply:
I like it. I'll change it.

Reply:
Great, just a reminder for the same addition to the "P" parameter above it too.